firmware/applications/SignBuddy/comms.c

```diff
@@ -0,0 +1,174 @@
+#include "comms.h"
+
+#include "ble_uart.h"
+#include "board.h"
+#include "common.h"
+#include "crc.h"
+#include "fifo.h"
+#include "logger.h"
+#include "pb_encode.h"
+#include "SignBuddy.pb.h"
+
+#include <stdlib.h>
+
+#define COMMS_TX_BUFFER_SIZE    512
+
+#define SYNC_BEGIN              0x16
+#define SYNC_END                0x27
+
+#define MID_SAMPLE              0x01
+
+#define PACKET_SIZE             sizeof(packet_header_t) + s.packet_header.length * sizeof(uint8_t) + \
+  sizeof(packet_footer_t)
+
+typedef struct __attribute__((__packed__)) {
+  uint8_t sync_begin;
+  uint8_t mid;
+  uint8_t length;
+} packet_header_t;
+
+typedef struct __attribute__((__packed__)) {
+  uint32_t crc;
+  uint8_t sync_end;
+} packet_footer_t;
+
+typedef struct {
+  TaskHandle_t      task_handle;
+  packet_header_t   packet_header;
+  packet_footer_t   packet_footer;
+  Sample            sample;
+  uint8_t           sample_buffer[Sample_size];
+  uint8_t          *packed_sample_ptr;
+  volatile uint8_t  sample_ready;
+  volatile uint8_t  packet_ready;
+  uint8_t           tx_buffer[COMMS_TX_BUFFER_SIZE];
+  fifo_t            tx_fifo;
+  SemaphoreHandle_t sample_mutex;
+} state_t;
+
+static state_t s;
+
+static void hw_init(void)
+{
+  ble_uart_init();
+}
+
+static void rx()
+{
+  DISABLE_IRQ();
+  uint8_t cmd = LL_LPUART_ReceiveData8(BLE_UART);
+  ENABLE_IRQ();
+
+  switch (cmd) {
+  case 0:
+    LOG_DEBUG("comms: cmd 1\r\n");
+    break;
+
+  case 1:
+    LOG_DEBUG("comms: cmd 2\r\n");
+    break;
+
+  case 2:
+    LOG_DEBUG("comms: cmd 3\r\n");
+    break;
+
+  default:
+    LOG_DEBUG("comms: unknown cmd\r\n");
+    break;
+  }
+}
+
+static void ingest_packet()
+{
+  fifo_push(&s.tx_fifo, (uint8_t *) &s.packet_header, sizeof(packet_header_t));
+  fifo_push(&s.tx_fifo, s.packed_sample_ptr, s.packet_header.length);
+  fifo_push(&s.tx_fifo, (uint8_t *) &s.packet_footer, sizeof(packet_footer_t));
+  free(s.packed_sample_ptr);
+  s.packet_ready = 0;
+}
+
+static void packetize_sample()
+{
+  pb_ostream_t stream = pb_ostream_from_buffer((pb_byte_t *) s.sample_buffer, Sample_size);
+
+  pb_encode(&stream, &Sample_msg, &s.sample);
+  s.sample_ready = 0;
+  s.packet_header.length = stream.bytes_written;
+  s.packed_sample_ptr = malloc(s.packet_header.length);
+  s.packet_header.mid = MID_SAMPLE;
+  memcpy(s.packed_sample_ptr, &s.sample_buffer, s.packet_header.length);
```

Comment:
i don't understand the point of copying the data into this `s.packed_sample_ptr`

Reply:
it seems like you just allocate it, copy the data into it, and then push to the fifo from it, when the other buffer is not being changed

Reply:
Yeah I'll get rid of the extra pointer and just push to the fifo from the buffer